Add first draft of OpenPBS power limit plugin

avilcheslopez opened the pull request:
Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

- Relates to #2805 
- Fixes #2807

avilcheslopez committed:
Add first draft of OpenPBS power limit plugin

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez commented on service/integration/open_pbs/geopm_power_limit.py:
```diff
@@ -0,0 +1,145 @@
+from geopmdpy import pio
+
+
+_SAVED_CONTROLS_FILE = "run/geopm-service/SAVE_FILES/saved_controls.json"
```
Haven't decided on location and file name. We could put the file in the same location as the other save files. If so, perhaps the file could be named something like: "power-limit-save-control.json".

Alternatively, if we don't want to mix it with the other files, we could put it in a separate folder, for example: "/run/geopm-pbs-hooks/SAVE_FILES" or "/run/geopm-extensions/SAVE_FILES"

My preference would be to put it in a separate folder since the files are managed by separate modules/mechanisms.

avilcheslopez replied:
Is there a "run" directory for open_pbs ("/run/open_pbs")? If not, "geopm-pbs-hooks" would be fine. First look for other examples of PBS hooks.

avilcheslopez commented on service/integration/open_pbs/geopm_power_limit.py:
```diff
@@ -0,0 +1,145 @@
+
+_SAVED_CONTROLS_FILE = "run/geopm-service/SAVE_FILES/saved_controls.json"
+_POWER_LIMIT_RESOURCE = "geopm-node-power-limit"
+_SIGNAL = "GPU_CORE_FREQUENCY_MIN_CONTROL"
+_CONTROL = "GPU_CORE_FREQUENCY_MIN_CONTROL"
```
Used for testing. Will update with actual signals/controls soon.

avilcheslopez committed:
Use power limit controls in OpenPBS hook

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add OpenPBS hook installation script

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add OpenPBS README file

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Handle requeued job in OpenPBS hook

On permanent error conditions, reject job and prevent it from being
requeued.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Securely create OpenPBS hook saved controls directory

If directory doesn't exist, securely create it using geopmdpy's
system_files module.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add unit tests for OpenPBS power limit hook prologue

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add unit tests for OpenPBS power limit hook epilogue

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add license headers to OpenPBS power limit hook files

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add OpenPBS files to dist

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez commented on service/integration/open_pbs/geopm_power_limit.py:
```diff
@@ -0,0 +1,163 @@
+    if not power_limit_requested:
+        if os.path.exists(_SAVED_CONTROLS_FILE):
+            restore_controls_from_file(_SAVED_CONTROLS_FILE)
+        e.accept()
```
Add a return after calling accept()?

avilcheslopez commented on service/integration/open_pbs/geopm_power_limit.py:
```diff
@@ -0,0 +1,163 @@
+    elif event_type == pbs.EXECJOB_EPILOGUE:
+        do_power_limit_epilogue()
+    else:
+        reject_event("Power limit hook incorrectly configured!")
```
Put this in a function to facilitate unit testing

avilcheslopez commented on service/integration/open_pbs/geopm_power_limit.py:
```diff
@@ -0,0 +1,163 @@
+
+    power_limit_requested = False
+    try:
+        power_limit_str = server_job.Resource_List[_POWER_LIMIT_RESOURCE]
```
Does PBS do any accounting/management of the resource for you?

avilcheslopez commented on service/integration/open_pbs/geopm_power_limit.py:
```diff
@@ -0,0 +1,163 @@
+    read_controls(current_settings)
+    system_files.secure_make_dirs(_SAVED_CONTROLS_PATH)
+    save_controls_to_file(_SAVED_CONTROLS_FILE, current_settings)
+    _power_limit_control["setting"] = power_limit
+    write_controls(_controls)
```
Do you need to modify the global instances of `_power_limit_control`, `_controls`? -> Not really.

avilcheslopez commented on service/integration/open_pbs/geopm_power_limit.py:
```diff
@@ -0,0 +1,163 @@
+    pbs.logmsg(pbs.LOG_DEBUG, f"Requested power limit: {power_limit}")
+    current_settings = _controls.copy()
+    read_controls(current_settings)
+    system_files.secure_make_dirs(_SAVED_CONTROLS_PATH)
```
We should have a mechanism for removing this directory. Leverage installation script.

avilcheslopez commented on service/integration/open_pbs/geopm_install_pbs_power_limit.sh:
```diff
@@ -0,0 +1,28 @@
+out=`qmgr -c "list resource" | grep "$RESOURCE"`
+if [ -z "$out" ]; then
+	echo "Creating $RESOURCE resource..."
+	qmgr -c "create resource $RESOURCE type=float" || exit 1
```
Note: the interface provided by PlatformIO takes a float for controls. Update the documentation to reflect that.

avilcheslopez commented on service/integration/open_pbs/geopm_install_pbs_power_limit.sh:
```diff
@@ -0,0 +1,28 @@
+#  Copyright (c) 2015 - 2023, Intel Corporation
+#  SPDX-License-Identifier: BSD-3-Clause
+#
+
```
Allow passing a "--remove" option to uninstall the hook, remove the "save file" directory.

avilcheslopez committed:
Misc. fixes for OpenPBS power limit hook

- Update documentation to reflect power limit may be a float
- Add unit tests for hook main body
- Rename saved controls file
- Add option to uninstall hook in installation script

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Make OpenPBS hook tests run with make check

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Fix OpenPBS hook test error

Mock out cffi

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Print logs for OpenPBS hook tests

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Fix OpenPBS hook test error

Some python modules were not being found

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Include geopm_openpbs_test.sh in dist

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

bgeltz commented on service/integration/open_pbs/README.rst:
```diff
@@ -0,0 +1,84 @@
+- geopmdpy (and libgeopmd) with LevelZero support on nodes where the power
+  limit feature is needed
+- This feature requires platform vendor HW support. In particular, it requires
+  HW support for certain MSRs: MSR_PLATFORM_LIMIT and
+  MSR_PLATFORM_ENGERGY_COUNTER
```
```suggestion
  HW support for certain MSRs: MSR::PLATFORM_POWER_LIMIT and
  MSR::PLATFORM_ENERGY_STATUS
```

bgeltz replied:
Actually, I'm not sure why you're calling out ```MSR::PLATFORM_ENERGY_STATUS``` here.  I don't see it being used in the hook.

avilcheslopez replied:
Fixed the naming.

The PLATFORM_ENERGY_STATUS is not needed directly by the hook, but it's an indirect dependency, as the IOGroup uses it internally.

bgeltz commented on service/integration/open_pbs/geopm_power_limit.py:
```diff
@@ -0,0 +1,168 @@
+
+
+def do_power_limit_prologue():
+    # pbs.logmsg(pbs.LOG_DEBUG, "Executing geopm_power_limit prologue")
+
```
```suggestion
```

bgeltz commented on service/integration/open_pbs/geopm_power_limit.py:
```diff
@@ -0,0 +1,168 @@
+
+
+def do_power_limit_epilogue():
+    # pbs.logmsg(pbs.LOG_DEBUG, "Executing geopm_power_limit epilogue")
+
```
```suggestion
```

bgeltz commented on service/integration/open_pbs/geopm_power_limit.py:
```diff
@@ -0,0 +1,168 @@
+         "domain_type": "board",
+         "domain_idx": 0,
+         "setting": 1}
+    ]
```
I'm wondering if we need to add handling for checking ```MSR::PLATFORM_POWER_LIMIT:LOCK``` to ensure these settings won't silently be ignored...

bgeltz replied:
In looking through the code here, Lowren and I proved to ourselves that when write_control() is called, the lock bit for the PLATFORM RAPL MSR will be checked and if set it will throw.  That exception will bubble out to this Python code, and should call into reject_event().

bgeltz commented on service/integration/open_pbs/README.rst:
```diff
@@ -0,0 +1,84 @@
+The GEOPM power limit hook requires:
+
+- OpenPBS
+- geopmdpy (and libgeopmd) with LevelZero support on nodes where the power
+  limit feature is needed
```
```suggestion
- geopmdpy (and libgeopmd) on nodes where the power limit feature is needed
```
Do we really need level zero?  All I see in the hook are MSRs.

avilcheslopez replied:
Good point. This was confusion on my part. Will fix.

bgeltz commented on service/integration/open_pbs/geopm_power_limit.py:
```diff
@@ -0,0 +1,168 @@
+        power_limit_str = server_job.Resource_List[_POWER_LIMIT_RESOURCE]
+        power_limit_requested = bool(power_limit_str)
+    except KeyError:
+        # power_limit_requested = False
```
```suggestion
```

bgeltz commented on service/integration/open_pbs/geopm_power_limit.py:
```diff
@@ -0,0 +1,168 @@
+        return
+
+    power_limit = resource_to_float(_POWER_LIMIT_RESOURCE, power_limit_str)
+    pbs.logmsg(pbs.LOG_DEBUG, f"Requested power limit: {power_limit}")
```
Is it evident in the PBS logs that this message is emitted by this plugin?  i.e., it would be redundant to prefix it with ```"<geopm>"```?

avilcheslopez replied:
Good point. There is an indicator the message comes from a hook, but the message doesn't tag/label/prefix the message with the hook name. I'll add our `<geopm>` tag.

bgeltz commented on service/integration/open_pbs/geopm_power_limit.py:
```diff
@@ -0,0 +1,168 @@
+        {"name": "MSR::PLATFORM_POWER_LIMIT::PL1_TIME_WINDOW",
+         "domain_type": "board",
+         "domain_idx": 0,
+         "setting": 0.013},
```
Is this different than the boot time default?  Wondering if we should just leave it alone.

avilcheslopez replied:
It is. These are the recommended settings Lowren provided (they come from the SDM). Lowren observed better performance vs. using the default.

bgeltz reviewed (changes requested):
Looks pretty good.  A few questions and fixups.

avilcheslopez committed:
Minor fixes/clean-up in OpenPBS hook

- Documentation fixes
- Removed unneeded code comments in hook
- Prefixed hook log messages with <geopm>

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

bgeltz commented on service/integration/open_pbs/geopm_power_limit.py:
```diff
@@ -0,0 +1,163 @@
+        {"name": "MSR::PLATFORM_POWER_LIMIT::PL1_TIME_WINDOW",
+         "domain_type": "board",
+         "domain_idx": 0,
+         "setting": 0.013},
```
```suggestion
         "setting": 0.013}, # SDM Vol. 4. Table 2.39 - Recommends 0xD = 13
```

bgeltz committed:
Review feedback / Spelling fixes

Signed-off-by: Lowren Lawson <lowren.h.lawson@intel.com>
Co-authored-by: Brad Geltz <brad.geltz@intel.com>
